### PR TITLE
Integration test to defend subframe tutorial (and bugfix for subframes of attached collision objects)

### PR DESCRIPTION
**Pull request by tylerjw**
### Description

Issue: #1733
This PR is an integration test for the subframe tutorial that tests that the tutorial continues to work.

Future work:
* test orientation

Thank you @jaas14 for helping me write this.

### CODE REVIEW

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 1]
```diff
@@ -0,0 +1,299 @@
+/*********************************************************************
+* Software License Agreement (BSD License)
+*
+*  Copyright (c) 2019, Felix von Drigalski
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: I think this can be changed to you

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+*  modification, are permitted provided that the following conditions
+*  are met:
+*
+*   * Redistributions of source code must retain the above copyright
+*     notice, this list of conditions and the following disclaimer.
+*   * Redistributions in binary form must reproduce the above
+*     copyright notice, this list of conditions and the following
+*     disclaimer in the documentation and/or other materials provided
+*     with the distribution.
+*   * Neither the name of OMRON SINIC X nor the names of its
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: I think this can be changed to you

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+#include <gtest/gtest.h>
+
+// MoveIt
+#include <moveit/planning_scene_interface/planning_scene_interface.h>
+#include <moveit/move_group_interface/move_group_interface.h>
+
+// TF2
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+bool moveToCartPose(geometry_msgs::PoseStamped pose, moveit::planning_interface::MoveGroupInterface& group,
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: Add function comment header

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+  ROS_INFO_STREAM(pose.pose.position.x << ", " << pose.pose.position.y << ", " << pose.pose.position.z);
+  moveit::planning_interface::MoveGroupInterface::Plan myplan;
+  if (group.plan(myplan) && group.execute(myplan))
+    return true;
+
+  ROS_WARN("Failed to perform motion.");
+  return false;
+}
+
+void spawnCollisionObjects(moveit::planning_interface::PlanningSceneInterface& planning_scene_interface)
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: Add function comment header

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+  cylinder.subframe_poses[0].orientation = tf2::toMsg(orientation);
+
+  // Lastly, the objects are published to the PlanningScene. In this tutorial, we publish a box and a cylinder.
+  box.operation = moveit_msgs::CollisionObject::ADD;
+  cylinder.operation = moveit_msgs::CollisionObject::ADD;
+  std::vector<moveit_msgs::CollisionObject> collision_objects = { box, cylinder };
+  planning_scene_interface.applyCollisionObjects(collision_objects);
+}
+
+void poseMsgToEigen(const geometry_msgs::Pose& m, Eigen::Affine3d& e)
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: do not use one letter variable names

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+TEST(TestPlanUsingSubframes, SubframesTests)
+{
+  ros::NodeHandle nh;
+  ros::AsyncSpinner spinner(1);
+  spinner.start();
+
+  geometry_msgs::Pose tip_in_hand_msg, hand_in_world_msg, tip_in_world_msg;
+  Eigen::Affine3d tip_in_hand, tip_in_world, hand_in_word;
+
+  ros::WallDuration(1.0).sleep();
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: this should be removed, or a really detailed comment explain why its needed

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+  temp_pose_stamped.header.frame_id = "box/bottom";
+
+  // When constructing the target pose for the robot, we multiply the quaternions to get the
+  // target orientation and convert the result to a ``geometry_msgs/orientation`` message.
+  orientation_1.setRPY(-(90.0 / 180.0 * M_PI), 0, 0);
+  target_orientation = flip_around_y * orientation_1;
+  temp_pose_stamped.pose.orientation = tf2::toMsg(target_orientation);
+  temp_pose_stamped.pose.position.z = 0.01;
+  moveToCartPose(temp_pose_stamped, group, "cylinder/tip");
+  testNearSubframe("bottom", &planning_scene_interface, &group);
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: When using tests in a sub-function, use SCOPED_TRACE so you can easily track down where the error is coming from:

```
  {
    SCOPED_TRACE("Initial orientation");
    ....
    testNearSubframe(...);
  }
```

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+  planning_scene_interface.applyCollisionObjects(collision_objects);
+}
+
+void poseMsgToEigen(const geometry_msgs::Pose& m, Eigen::Affine3d& e)
+{
+  e = Eigen::Translation3d(m.position.x, m.position.y, m.position.z) *
+      Eigen::Quaterniond(m.orientation.w, m.orientation.x, m.orientation.y, m.orientation.z);
+}
+
+void testNearSubframe(const std::string& subframe_name,
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: add function comment

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+                      moveit::planning_interface::PlanningSceneInterface* planning_scene_interface,
+                      moveit::planning_interface::MoveGroupInterface* group)
+{
+  geometry_msgs::Pose tip_in_hand_msg, hand_in_world_msg, tip_in_world_msg, box_subframe_msg;
+  Eigen::Affine3d tip_in_hand, tip_in_world, hand_in_word, box_subframe_in_world;
+
+  auto attached_objects = planning_scene_interface->getAttachedObjects();
+  tip_in_hand_msg = attached_objects["cylinder"].object.subframe_poses[0];
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: EXPECT_EQ(attached_objects["cylinder"].object.subframe_poses.size(), 1)

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+  hand_in_world_msg = group->getCurrentPose("panda_hand").pose;
+  poseMsgToEigen(tip_in_hand_msg, tip_in_hand);
+  poseMsgToEigen(hand_in_world_msg, hand_in_word);
+  tip_in_world = hand_in_word * tip_in_hand;
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: check that tip_in_world has no NaNs

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+
+  auto box = planning_scene_interface->getObjects()["box"];
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: check that an object was actually found named box

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+  auto box_subframe_names = box.subframe_names;
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: We discourage usage of auto in MoveIt, as it hurts readability. Is this a string? 

> However, auto is not an expression highly readable. Please explicitly specify the variable type, if it doesn’t become clear immediately from the context.

Comment (davetcoleman) [reply moveit_ros/planning_interface/test/plan_using_subframes.cpp]: https://moveit.ros.org/documentation/contributing/code/

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+  bool found_box_subframe = false;
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: move this for loop into a helper sub function

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+      return true;
+    }
+  }
+  return false;
+}
+
+// Test method to test if the cylinder tip at near the box subrame_name meantioned
+void testAtSubframe(const std::string& subframe_name,
+                    moveit::planning_interface::PlanningSceneInterface* planning_scene_interface,
+                    moveit::planning_interface::MoveGroupInterface* group)
```
Comment (felixvd) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: I would add `double tolerance = 0.1` and use that at the end of the function (although that tolerance is so huge it's almost meaningless. I'd set it to at least 0.01)

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+  // When constructing the target pose for the robot, we multiply the quaternions to get the
+  // target orientation and convert the result to a ``geometry_msgs/orientation`` message.
+  orientation_1.setRPY(-(90.0 / 180.0 * M_PI), 0, 0);
+  target_orientation = flip_around_y * orientation_1;
+  temp_pose_stamped.pose.orientation = tf2::toMsg(target_orientation);
+  temp_pose_stamped.pose.position.z = 0.01;
+  moveToCartPose(temp_pose_stamped, group, "cylinder/tip");
+  testAtSubframe("bottom", &planning_scene_interface, &group);
+
+  // The command "2" moves the cylinder tip to the top of the box (the right side in the top animation).
```
Comment (felixvd) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]:
```suggestion
```

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+  ROS_INFO_STREAM("Moving to top of box with cylinder tip");
+  temp_pose_stamped.header.frame_id = "box/top";
+  orientation_1.setRPY((90.0 / 180.0 * M_PI), 0, 0);
+  target_orientation = flip_around_y * orientation_1;
+  temp_pose_stamped.pose.orientation = tf2::toMsg(target_orientation);
+  temp_pose_stamped.pose.position.z = 0.01;
+  moveToCartPose(temp_pose_stamped, group, "cylinder/tip");
```
Comment (felixvd) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]:
```suggestion
  ASSERT_TRUE(moveToCartPose(temp_pose_stamped, group, "cylinder/tip"));
```
Since this checks for planning & execution success, it makes sense to fail the test if false is returned.

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+    {
+      *index = i;
+      return true;
+    }
+  }
+  return false;
+}
+
+// Test method to test if the cylinder tip at near the box subrame_name meantioned
+void testAtSubframe(const std::string& subframe_name,
```
Comment (felixvd) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: I am always for much more verbose function names. I would call this "isCylinderTipAtBoxSubframe" or something similar.

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp (hunk 16 above)
Comment (felixvd) [reply moveit_ros/planning_interface/test/plan_using_subframes.cpp]: Same for the next calls of this further down

File: moveit_ros/planning_interface/test/plan_using_subframes.cpp [hunk 18]
```diff
@@ -0,0 +1,323 @@
...
+  Eigen::Affine3d tip_in_hand, tip_in_world, hand_in_word, box_subframe_in_world;
+
+  std::map<std::string, moveit_msgs::AttachedCollisionObject> attached_objects =
+      planning_scene_interface->getAttachedObjects();
+  EXPECT_EQ(attached_objects["cylinder"].object.subframe_poses.size(), std::size_t(1));
+  tip_in_hand_msg = attached_objects["cylinder"].object.subframe_poses[0];
+  hand_in_world_msg = group->getCurrentPose("panda_hand").pose;
+  tip_in_hand_msg.position.x += 0.01;
+  poseMsgToEigen(tip_in_hand, tip_in_hand_msg);
+  poseMsgToEigen(hand_in_word, hand_in_world_msg);
```
Comment (felixvd) [inline moveit_ros/planning_interface/test/plan_using_subframes.cpp]: word --> world

**Comment by tylerjw** [comment]
@felixvd I'm not sure how this passed locally for me.  It looks like even with the huge margin (0.1) I didn't get close enough with the math.  I'll revisit this and address your feedback this next week when I have time.  Thank you for the review.

**Comment by tylerjw** [comment]
@felixvd I'm sorry this took me a while to get around to.  I responded to your feedback and I believe this is now in a much better state.  Please review (assuming it passes CI).

**Comment by tylerjw** [comment]
This doesn't pass on kinetic.  Is there a chance subframes isn't backported or something similar?

File: moveit_ros/planning_interface/test/subframes_test.cpp [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+#include <gtest/gtest.h>
+
+// MoveIt
+#include <moveit/planning_scene_interface/planning_scene_interface.h>
+#include <moveit/move_group_interface/move_group_interface.h>
+
+// TF2
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+constexpr double kEpsilon = 1e-3;
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: The k prefix is a historical oddity and means that it is a constant.  Here is a SO post about it: https://stackoverflow.com/questions/5016622/where-does-the-k-prefix-for-constants-come-from

As to the style, the k prefix with Camel case is the standard in the google style guide: https://google.github.io/styleguide/cppguide.html#Constant_Names

File: moveit_ros/planning_interface/test/subframes_test.cpp [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+  Eigen::Affine3d tip_in_hand, hand_in_world;
+
+  std::map<std::string, moveit_msgs::AttachedCollisionObject> attached_objects =
+      planning_scene_interface->getAttachedObjects();
+  if (attached_objects.find("cylinder") == attached_objects.end())
+  {
+    ROS_ERROR_NAMED(log_name, "No cylinder object");
+    return false;
+  }
+  if (attached_objects["cylinder"].object.subframe_poses.size() != std::size_t(1))
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: That somehow the number of subframes changed.  I removed it anyway due to the other changes.

File: moveit_ros/planning_interface/test/subframes_test.cpp [hunk 21]
```diff
@@ -0,0 +1,366 @@
...
+
+// Helper function to convert pose message to eigen Affine3d
+void poseMsgToEigen(Eigen::Affine3d& left, const geometry_msgs::Pose& right)
+{
+  left = Eigen::Translation3d(right.position.x, right.position.y, right.position.z) *
+         Eigen::Quaterniond(right.orientation.w, right.orientation.x, right.orientation.y, right.orientation.z);
+}
+
+// Helper function for finding the index of a value in a list of strings
+bool findStringIndex(int* index, const std::vector<std::string>& list, const std::string& value)
```
[the remark this answers is not in the document]
Comment (tylerjw) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: I don't need it with the `getFrameTransform` method.

File: moveit_ros/planning_interface/test/subframes_test.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (davetcoleman) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: @tylerjw is used to Google style code bases, but unfortuantly that's not the standard we're using for MoveIt. Should be EPSILON or something like POSE_EPSILON

File: moveit_ros/planning_interface/test/subframes_test.cpp (hunk 20 above)
[the remark this answers is not in the document]
Comment (davetcoleman) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: More tests, even if seemingly obvious, aren't bad IMHO

**Comment by davetcoleman** [comment]
@tylerjw let's get this merged in :-)

**Comment by tylerjw** [comment]
@mlautman I got the psm to work as expected.  Now I'm actually having some weird issue with subframes.  Here is the debug output, the values in cylinder_tip look wrong, however box_subframe seem reasonable:

```
panda_hand: 0.306957, 0.0899721, 0.460283
box_subframe: 0.307, 0.05, 0.34027
cylinder: 0.307055, 0.0899991, 0.340283
cylinder_tip: 0.03, 1.68051e-18, 0.12
```

panda_subframe

**Comment by tylerjw** [comment]
For more debugging here is the planning scene message representation of `cylinder`:

```
    attached_collision_objects: 
      - 
        link_name: "panda_hand"
        object: 
          header: 
            seq: 0
            stamp: 
              secs: 0
              nsecs:         0
            frame_id: "panda_hand"
          id: "cylinder"
          type: 
            key: ''
            db: ''
          primitives: 
            - 
              type: 3
              dimensions: [0.06, 0.005]
          primitive_poses: 
            - 
              position: 
                x: 0.0
                y: 0.0
                z: 0.12
              orientation: 
                x: -1.96644880614e-17
                y: 0.707106781187
                z: -1.96644880626e-17
                w: 0.707106781187
          meshes: []
          mesh_poses: []
          planes: []
          plane_poses: []
          subframe_names: [tip]
          subframe_poses: 
            - 
              position: 
                x: 0.03
                y: 1.68051336735e-18
                z: 0.12
              orientation: 
                x: -1.96644880614e-17
                y: 0.707106781187
                z: -1.96644880626e-17
                w: 0.707106781187
          operation: 0
        touch_links: []
        detach_posture: 
          header: 
            seq: 0
            stamp: 
              secs: 0
              nsecs:         0
            frame_id: ''
          joint_names: []
          points: []
        weight: 0.0
    is_diff: False
```

**Comment by tylerjw** [comment]
Based on what I posted above, I think there might be a bug in how `getFrameTransform` gets the transform for subframes of attached collision objects.  You'll notice the "tip" position reported is relative to the hand and not the world.

**Comment by tylerjw** [comment]
Before you ask, I rebased on master and pushed.  Did not change the results.

**Comment by tylerjw** [comment]
To run the test use this command after sourcing your workspace and building with the `--no-install` option:

```rostest moveit_ros_planning_interface subframes_test.test```

**Comment by tylerjw** [comment]
@rhaschke thank you for fixing the bug for this.  Assuming this passes travis I believe this is ready for review again.  @felixvd

**Comment by tylerjw** [comment]
Sadness, this failed on line 259 of the test for clang only.  Any ideas why that would be?

Here is a link to the test that failed: https://travis-ci.org/github/ros-planning/moveit/jobs/667472631?utm_medium=notification&utm_source=github_status

**Comment by tylerjw** [comment]
On the plus side this increases code coverage in the whole repo by 1.57% https://codecov.io/gh/ros-planning/moveit/pull/1757/diff

File: moveit_ros/planning_interface/test/subframes_test.cpp [hunk 22]
```diff
@@ -86,7 +86,7 @@ void spawnCollisionObjects(moveit::planning_interface::PlanningSceneInterface& p
 {
   const std::string log_name = "spawn_collision_objects";
   double z_offset_box = .25;  // The z-axis points away from the gripper
-  double z_offset_cylinder = .1;
+  double z_offset_cylinder = .12;
```
Comment (rhaschke) [inline moveit_ros/planning_interface/test/subframes_test.cpp]: Why did you change that value back?
For me, this change was required to have the cylinder centered between the jaws.

File: moveit_ros/planning_interface/test/subframes_test.cpp [hunk 23]
```diff
@@ -193,11 +184,7 @@ TEST(TestPlanUsingSubframes, SubframesTests)
   geometry_msgs::Pose tip_in_hand_msg, hand_in_world_msg, tip_in_world_msg;
```
Comment (rhaschke) [inline moveit_ros/planning_interface/test/subframes_test.cpp]: The variables in the following two lines are unused. Please removed them.

File: moveit_ros/planning_interface/test/subframes_test.cpp [hunk 24]
```diff
@@ -199,50 +199,108 @@ TEST(TestPlanUsingSubframes, SubframesTests)
   tf2::Quaternion target_orientation;
   target_orientation.setRPY(0, 180.0 / 180.0 * M_PI, 90.0 / 180.0 * M_PI);
   geometry_msgs::PoseStamped target_pose_stamped;
-  target_pose_stamped.header.frame_id = "box/bottom";
   target_pose_stamped.pose.orientation = tf2::toMsg(target_orientation);
   target_pose_stamped.pose.position.z = Z_OFFSET;
 
   ROS_INFO_STREAM_NAMED(log_name, "Moving to bottom of box with cylinder tip");
+  target_pose_stamped.header.frame_id = "box/bottom";
+  ASSERT_TRUE(moveToCartPose(target_pose_stamped, group, "cylinder/tip"));
```
Comment (rhaschke) [inline moveit_ros/planning_interface/test/subframes_test.cpp]: The following block is the same for all subsequent checks. Please factor this out into a function:

```c++
void moveAndValidate(const geometry_msgs::PoseStamped& pose, moveit::planning_interface::MoveGroupInterface& group,
                     const std::string& eef_link, const planning_scene_monitor::PlanningSceneMonitorPtr& psm)
{
  ASSERT_TRUE(moveToCartPose(pose, group, eef_link));
  psm->requestPlanningSceneState();
  planning_scene_monitor::LockedPlanningSceneRO planning_scene(psm);

  // get the tip and box subframe locations in world
  Eigen::Isometry3d eef = planning_scene->getFrameTransform(eef_link);
  Eigen::Isometry3d box_subframe = planning_scene->getFrameTransform(pose.header.frame_id);
  Eigen::Isometry3d target_pose;
  tf::poseMsgToEigen(pose.pose, target_pose);

  // expect that they are identical
  EXPECT_TRUE(eef.isApprox(box_subframe * target_pose, EPSILON)) << "box frame: \n"
                                                                 << box_subframe.matrix() << "\ncylinder frame: \n"
                                                                 << eef.matrix();
}
```

Comment (rhaschke) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: I'm not sure though that we should include all these tests at all. They essentially check exactly the same functionality and thus just waste CI time.

Comment (rhaschke) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: You can find my suggested changes here: https://github.com/ubi-agni/moveit/tree/p/tylerjw/subframes_test

**Comment by rhaschke** [comment]
Rerunning Travis, I have got a failure for clang-tidy only (which is fixed in my version already).
So, I think you have seen a flaky test. This, also, suggests to drop all the other targets and only keep `box/bottom`.

**Comment by rhaschke** [comment]
Fixes 1974.

File: moveit_ros/planning_interface/test/subframes_test.cpp (hunk 22 above)
Comment (tylerjw) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: When I tested locally this change caused it to fail the motion plan because it was in collision.  I posted the log output in your PR.

Comment (rhaschke) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: Do you use an outdated panda_moveit_config or franka_description locally?
As https://travis-ci.com/github/ubi-agni/moveit/jobs/305268241 shows, the tests pass with 0.1.

File: moveit_ros/planning_interface/test/subframes_test.cpp (hunk 24 above)
Comment (tylerjw) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: I agree about only checking one of them and that this is just extra redundancy.

File: moveit_ros/planning_interface/test/subframes_test.cpp (hunk 22 above)
Comment (tylerjw) [reply moveit_ros/planning_interface/test/subframes_test.cpp]:
```ros.moveit_core.collision_detection.fcl: Found a contact between 'cylinder' (type 'Robot attached') and 'panda_rightfinger' (type 'Robot link'), which constitutes a collision. Contact information is not stored.
ros.moveit_core.collision_detection.fcl: Collision checking is considered complete (collision was found and 0 contacts are stored)
ros.moveit_ros_planning: Start state appears to be in collision with respect to group panda_arm
ros.moveit_ros_planning: Unable to find a valid state nearby the start state (using jiggle fraction of 0.050000 and 100 sampling attempts). Passing the original planning request to the planner.
ros.moveit_planners_ompl.planning_context_manager: Using 'JointModel' parameterization for solving problem
ros.moveit_planners_ompl.planning_context_manager: Creating new planning context
ros.moveit_planners_ompl.model_based_planning_context: panda_arm: Setting planning volume (affects SE2 & SE3 joints only) to x = [-5.000000, 5.000000], y = [-5.000000, 5.000000], z = [-5.000000, 5.000000]
ros.moveit_core.kinematic_constraints: The desired rotation matrix for link 'panda_hand' in frame world is:
 0.000398163           -1  7.40105e-16
          -1 -0.000398163  -6.9279e-12
  6.9279e-12  2.01833e-15           -1
ros.moveit_core.kinematic_constraints: The desired rotation matrix for link 'panda_hand' in frame world is:
 0.000398163           -1  7.40105e-16
          -1 -0.000398163  -6.9279e-12
  6.9279e-12  2.01833e-15           -1
ros.moveit_planners_ompl.constrained_goal_sampler: Constructed a ConstrainedGoalSampler instance at address 0x7fd8700131d0
ros.moveit_planners_ompl.model_based_planning_context: Planner configuration 'panda_arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
ros.moveit_planners_ompl.planning_context_manager: panda_arm: New planning context is set.
ros.moveit_planners_ompl.model_based_planning_context: panda_arm: Solving the planning problem once...
ros.moveit_planners_ompl.ompl: panda_arm/panda_arm: Skipping invalid start state (invalid state)
ros.moveit_planners_ompl.ompl: panda_arm/panda_arm: Motion planning start tree could not be initialized!
ros.moveit_planners_ompl.ompl: No solution found after 0.000572 seconds
ros.moveit_planners_ompl.ompl: Goal sampling thread never did any work.
ros.moveit_planners_ompl.model_based_planning_context: There were 0 valid motions and 0 invalid motions.
ros.moveit_planners_ompl.model_based_planning_context: Unable to solve the planning problem
ros.moveit_ros_planning_interface.move_group_interface: Fail: ABORTED: No motion plan found. No execution attempted.
ros.moveit_ros_planning_interface: Failed to perform motion.
```

Tested again locally and this is what happens on mine when I change it to 0.1.  Note that in the tutorial this test is based off the value is 0.12.

**Comment by tylerjw** [comment]
@rhaschke I cherry-picked your latest change onto here and responded to your feedback about the extra steps in the test that will just waste CI's time.  I don't think factoring it out into a function is necessary now though so I put that back.  Please re-review.

File: moveit_ros/planning_interface/test/subframes_test.cpp (hunk 22 above)
Comment (tylerjw) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: > Do you use an outdated panda_moveit_config or franka_description locally?
> As https://travis-ci.com/github/ubi-agni/moveit/jobs/305268241 shows, the tests pass with 0.1.

If you go to the commit hash linked on that travis job it ran with 0.12.  See: https://github.com/ubi-agni/moveit/blob/a18d278014375764360c78626d53cdc0a0cfa14d/moveit_ros/planning_interface/test/subframes_test.cpp

Comment (tylerjw) [reply moveit_ros/planning_interface/test/subframes_test.cpp]: > Do you use an outdated panda_moveit_config or franka_description locally?

That was it!  I didn't realize there was a source install for a newer panda_moveit_config called out in the rosinstall.  Did that happen recently?  Anyway, it is now working for me locally at 0.1 and I updated it to be 0.1.

**Comment by tylerjw** [comment]
@rhaschke set it to 0.1 as I got it to work locally.  Ready for review.

**Comment by tylerjw** [comment]
Travis is failing to update the status on github but it passed, here is a link to the run off the latest commit: https://travis-ci.org/github/ros-planning/moveit/builds/667830874?utm_medium=notification&utm_source=github_status

**Comment by tylerjw** [comment]
rebased on master to see if that'd kick travis into working... didn't seem to work :/

**Comment by tylerjw** [comment]
It looks like this is a flaky test (as it failed on one of the other tests this time).  @rhaschke do you have any ideas on how to make this less flaky and yet still test the behavior of the tutorial and subframes?

**Comment by tylerjw** [comment]
Once again, travis doesn't seem to be syncing with github here, but this latest one passed on all tests.

**Comment by rhaschke** [comment]
We need to investigate this flaky test in more detail. The orientation should match as well!

**Comment by tylerjw** [comment]
Putting epsilon back to 1e-3 (1mm) I can get it to fail locally after running the test a dozen or so times.  Here is what the robot looks like when it failed:

local_fail

Here is the output from the console:
```
[ROSTEST]-----------------------------------------------------------------------

[moveit_ros_planning_interface.rosunit-subframes_test/SubframesTests][FAILURE]--
/home/tyler/workspace/ws_moveit/src/moveit/moveit_ros/planning_interface/test/subframes_test.cpp:184
Value of: eef.isApprox(box_subframe * target_pose, EPSILON)
  Actual: false
Expected: true
box frame: 
           1  7.41052e-17 -0.000398163        0.307
 0.000398163 -6.92801e-12            1         0.05
-2.68432e-15           -1 -6.92801e-12      0.34027
           0            0            0            1
cylinder frame: 
-0.000920403    -0.999999     0.001371     0.306928
-0.000729613  -0.00137033    -0.999999    0.0600662
    0.999999 -0.000921402 -0.000728351     0.340262
           0            0            0            1
--------------------------------------------------------------------------------
```

For comparison this is a success:
```
ros.moveit_ros_planning_interface.test_plan_using_subframes: box frame: 
           1  7.41052e-17 -0.000398163        0.307
 0.000398163 -6.92801e-12            1         0.05
-2.68432e-15           -1 -6.92801e-12      0.34027
           0            0            0            1
cylinder frame: 
 -0.00084591    -0.999999   0.00136466     0.306935
  0.00015513  -0.00136479    -0.999999    0.0599787
           1 -0.000845698  0.000156284     0.340268
           0            0            0            1
```

Can someone who knows more about the math of a rotation matrix see if they can understand what is happening wrong here? @AndyZe @rhaschke

**Comment by tylerjw** [comment]
I wrote a bit of python to compare the failing and passing difference between the box and cylinder subframes and this is the result:

```
Failing diff:
[[  1.00092040e+00   9.99999000e-01  -1.76916300e-03   7.20000000e-05]
 [  1.12777600e-03   1.37032999e-03   1.99999900e+00  -1.00662000e-02]
 [ -9.99999000e-01  -9.99078598e-01   7.28350993e-04   8.00000000e-06]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   0.00000000e+00]]
Passing diff:
[[  1.00084591e+00   9.99999000e-01  -1.76282300e-03   6.50000000e-05]
 [  2.43033000e-04   1.36478999e-03   1.99999900e+00  -9.97870000e-03]
 [ -1.00000000e+00  -9.99154302e-01  -1.56284007e-04   2.00000000e-06]
 [  0.00000000e+00   0.00000000e+00   0.00000000e+00   0.00000000e+00]]
```

**Comment by tylerjw** [comment]
My suggestion is we relax the requirement of epsilon to account for small differences in the ending pose from 1mm to 1cm.  I can't get it to fail locally at 1cm.

I don't think is actually a flaky test, I think we were just testing the result of a motion plan too finely.

@rhaschke

File: moveit_ros/planning_interface/test/subframes_test.cpp [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+
+    // get the tip and box subframe locations in world
+    Eigen::Isometry3d eef = planning_scene->getFrameTransform("cylinder/tip");
+    Eigen::Isometry3d box_subframe = planning_scene->getFrameTransform(target_pose_stamped.header.frame_id);
+    Eigen::Isometry3d target_pose;
+    tf::poseMsgToEigen(target_pose_stamped.pose, target_pose);
+
+    // expect that they are identical
+    std::stringstream ss;
+    ss << "box frame: \n" << box_subframe.matrix() << "\ncylinder frame: \n" << eef.matrix();
```
Comment (rhaschke) [inline moveit_ros/planning_interface/test/subframes_test.cpp]: We should print the correct transform for comparison.
```suggestion
    ss << "target frame: \n" << (box_subframe * target_pose).matrix() << "\ncylinder frame: \n" << eef.matrix();
```

**Comment by tylerjw** [comment]
done @rhaschke

**Comment by felixvd** [comment]
Thanks to both of you for cleaning up after this :)

**Comment by v4hn** [comment]
Congratulations @tylerjw on seeing this through to the merge! :)

**Comment by tylerjw** [comment]
@felixvd @v4hn @rhaschke thank you for the help getting this done.  I learned so much about the interfaces of moveit doing this.

**Comment by poweic** [comment]
Just stumbled on this. Should we add `eigen_conversions` to `CATKIN_DEPENDS` section in `catkin_package` in file [moveit_core/CMakeLists.txt](https://github.com/ros-planning/moveit/blob/master/moveit_core/CMakeLists.txt)?

```cmake
catkin_package(
  INCLUDE_DIRS
    ${THIS_PACKAGE_INCLUDE_DIRS}
  LIBRARIES
    moveit_exceptions
    # more here ...
    ${OCTOMAP_LIBRARIES}
  CATKIN_DEPENDS
    eigen_stl_containers
    eigen_conversions  # <=============== should we add this ???
    geometric_shapes
    geometry_msgs
    kdl_parser
    moveit_msgs
    octomap_msgs
    random_numbers
    sensor_msgs
    shape_msgs
    srdfdom
    std_msgs
    tf2_eigen
    tf2_geometry_msgs
    trajectory_msgs
    visualization_msgs
  DEPENDS
    Boost
    # more here ...
    ${BULLET_ENABLE}
    )
```

**Comment by rhaschke** [comment]
@poweic, what kind of issue did you observe?
Please file a new PR, if you think you have a fix for something.
CATKIN_DEPENS should list dependencies that are transitive, i.e. are required by downstream packages of moveit_core, e.g. because the eigen_conversion headers are included in our headers.
